File: leetcode/24/1820.maximum-number-of-accepted-invitations.cpp
#include <istream>
#include <iostream>
#include <string>
#include <vector>
#include <array>
#include <map>

using namespace std;

class Solution {
    public:
    bool bipartiteMatch(const vector<vector<int>>& grid, int u, vector<bool> visited, vector<int>& girls) {
        int m = grid.size();
        int n = grid[0].size();
        for (int v=0; v<n; v++) {
            if (grid[u][v] && !visited[v]) {
                visited[v] = true;
                if (girls[v]<0 || bipartiteMatch(grid,girls[v],visited,girls)) {
                    girls[v] = u;
                    return true;
                }
            }
        }
        return false;
    }

    int maximumInvitations(vector<vector<int>>& grid) {
        int matches = 0;
        int m = grid.size();
        int n = grid[0].size();
        vector<int> girls(n,-1);

        for (int u=0; u<m; u++) {
            vector<bool> visited(n,false);
            if (bipartiteMatch(grid,u,visited,girls)) {
                matches++;
            }
        }
        return matches;
    }
};

File: leetcode/24/1820.maximum-number-of-accepted-invitations.py
#
# @lc app=leetcode id=1820 lang=python3
#
# [1820] Maximum Number of Accepted Invitations
#
# @lc code=start
from collections import defaultdict

class Solution:
    def maximumInvitations(self, grid):
        boys = len(grid)
        girls = len(grid[0])
        kids = defaultdict(int)

        def dfs(boy, visit):
            for girl in range(girls):
                if grid[boy][girl] and girl not in visit:
                    visit.add(girl)

                    if girl not in kids or dfs(kids[girl],visit):
                        kids[girl] = boy
                        return True
                    
            return False

        for boy in range(boys):
            dfs(boy,set())

        return len(kids)
# @lc code=end
